### DISCUSSION TITLE
Add and compile new primal solver

### DISCUSSION THREAD

**Discussion by sodabryant** [mdolab/dafoam#248] [Bug]
@friedenhe Dear Dr.He, I tried to add a new primal solver named myDASimpleFoam, and here is my update process:

    . $HOME/dafoam/loadDAFoam.sh
    cd /home/lxl/dafoam/repos/dafoam-2.2.5/src/adjoint
    ./Allmake
    cd $HOME/dafoam/repos/dafoam-2.2.5
    ./Allmake
    source $HOME/dafoam/OpenFOAM/OpenFOAM-v1812-AD/etc/bashrc
    ./Allclean && ./Allmake 2> errorLog.txt
    pip install .

The compilation last a real long time, then no error have been found. But, when I use the new myDASimpleFoam to run the pitzDaily case, I encountered the following error:

-------------------------------------------------------------------------------
|                               DAFoam v2.2.5                                 |
-------------------------------------------------------------------------------
 
Traceback (most recent call last):
  File "runScript.py", line 153, in <module>
    DASolver = PYDAFOAM(options=daOptions, comm=gcomm)
  File "/home/lxl/dafoam/packages/miniconda3/lib/python3.6/site-packages/dafoam/pyDAFoam.py", line 635, in __init__
    self._initSolver()
  File "/home/lxl/dafoam/packages/miniconda3/lib/python3.6/site-packages/dafoam/pyDAFoam.py", line 2162, in _initSolver
    from .pyDASolverIncompressible import pyDASolvers
ImportError: dynamic module does not define module export function (PyInit_pyDASolverIncompressible)
-------------------------------------------------------------------------------

I have no idea about this and don't know how to operate correctly. I am wondering whether the compile process is correct. Could you please offer me some assistance?

**Comment by friedenhe** [comment]
Can you elaborate on what exactly have you changed/added to add a new solver? It is likely that you missed some steps.

**Reply by sodabryant** [reply]
Yes sir! Except the existing solver, like DAsimpleFoam, DARhoSimpleFoam, etc. I wrote a new solver named myDASimpleFoam. Indeed, I want to try more operations on the "alphaPorosity" field, that's why I made a new solver.  
And here gives my modification process:
1) in dafoam/repos/dafoam-2.2.5/src/adjoint/DAResidual, I add DAResidualMySimpleFoam.C and DAResidualMySimpleFoam.H.
2) in dafoam/repos/dafoam-2.2.5/src/adjoint/DASolver, I add myDASimpleFoam/MyDASimpleFoam.C and myDASimpleFoam/MyDASimpleFoam.H.
3) in dafoam/repos/dafoam-2.2.5/src/adjoint/DAStateInfo, I add DAStateInfoMySimpleFoam.C and DAStateInfoMySimpleFoam.H.
4) in dafoam/repos/dafoam-2.2.5/src/adjoint/Make/files_Incompressible, I add 3lines as follows:
        DASolver/myDASimpleFoam/myDASimpleFoam.C
        DAStateInfo/DAStateInfoMySimpleFoam.C
        DAResidual/DAResidualMySimpleFoam.C
5) in dafoam/repos/dafoam-2.2.5, compile it with the command:  ./Allmake
6) source $HOME/dafoam/OpenFOAM/OpenFOAM-v1812-AD/etc/bashrc
    ./Allclean && ./Allmake 2> errorLog.txt
7) pip install .
After the operations above, I modified the file runScript.py, then I got the error as showed above. So where's my mistake? 

Hope for your reply. Thank you！

**Reply by friedenhe** [reply]
The overall steps look right. Did you change the class names in the new files? Did you change the `TypeName();` in the added .H files accordingly?

**Reply by sodabryant** [reply]
Thanks for your reply, I tried three times and now it works, maybe I missed some details before. 
Regards!

**Comment by sodabryant** [comment]
@friedenhe Dear Dr. He, I observed a complete optimization process of the pitzDaily case, but what bothers me now is that I can't find any criteria or conditions for ending up the optimization. Of course, the process automatively ends up when it fulfill the max iteration step 50 I defined. But it must have some other conditions to operate, Especially, what if I want to end up the optimization when the whole variable field, like alphaPorosity, not change anymore. Can you give me some hints about this?
Thank you!

**Reply by friedenhe** [reply]
Which optimizer did you use? You can definitely increase the max number of iterations to 50+ in the runScript.py.

**Reply by sodabryant** [reply]
I use IPOPT as the optimizer for now, but it shows a relatively worse result comparing with the result you presented before.  Maybe it indeed needs more iterations? Here is the alphaPorosity field after 50 steps. 
![alpha 0050](https://user-images.githubusercontent.com/49938783/149612617-a44488bf-0ac5-4abc-9547-2ea9c9ea3a81.jpg)

**Reply by friedenhe** [reply]
We only tested this case with SNOPT. We tried IPOPT and it did not work well for us. You may want to tweak the IPOPT parameters to see if they can improve the optimization.

**Reply by friedenhe** [reply]
BTW: you can find the optimality from IPOPT. Check the DAFoam workshop slides and video.

**Comment by sodabryant** [comment]
@friedenhe Dear He, I noticed that in your paper, you said that ADflow is a compressible solver, and DAFoam is a incompressible solver. But I found that the rhoSimpleFoam is included in the DAFoam, actually, I want to ask that if I want to further research the compressible flow with rhoCentralFoam, we know that rhoCentralFoam is a density based solver, is it possible to add rhoCentralFoam into the current DAFoam? Which state variables and residual equations should I define? Could you please offer me some assistance?

**Comment by friedenhe** [comment]
We added a few new features after the paper, and DAFoam supports subsonic and transonic cases now. But it does not work for supersonic flow yet. rhoCentralFoam is very different from the solvers already differentiated in DAFoam because it uses LTS. So we haven't gotten a chance to think about how to best differentiate it yet..
